### DISCUSSION TITLE
hide individual attributes from more-info dropdown

### DISCUSSION THREAD

**Discussion by Mariusthvdb** [NemesisRE/kiosk-mode#144] [Ideas]
just a thought to consider please, given the fact this Kiosk mode  plugin can hide a great many things, but does no want to 'card-mod' stuff:

would it be feasible to hide individual attributes from the dropdown?

example: we see the new enum device-class add a list of possible values to the attributes. To me these are more 'backend' attributes, (just like `device_class` itself is, a filtered attribute) and I don't need/want them in the dashboard . Ive filed a FR/discussion in the Frontend repo, but figured it might suit this Kiosk mode plugin just as well.

It (more-info updating) is a part of my own custom-ui to be able to do so, and Ive even made a separate js script. However, that is not up to the modern Dashboard implementation, and has its downsides.

I was hoping Kiosk mode would be the best candidate for an addition doing so, in a perfect , smooth and modern way. Just like anything else Kiosk mode does

thanks for reading thus far...

**Comment by elchininet** [comment]
Hi @Mariusthvdb,
What dropdown are you referring when you mention "more-info dropdown"?
Some screenshots can help to identify which UI element you are referring to.
Regards

**Reply by Mariusthvdb** [reply]
sorry, I figured that was obvious, my pardon..

consider a weather condition showing the attributes dropdown
<img width="570" alt="Scherm­afbeelding 2023-12-04 om 10 07 10" src="https://github.com/NemesisRE/kiosk-mode/assets/33354141/5d61282b-7b54-4680-8a4c-bff63e6d1fce">


and then clicking that only reveals the enum values:

<img width="569" alt="Scherm­afbeelding 2023-12-04 om 10 07 17" src="https://github.com/NemesisRE/kiosk-mode/assets/33354141/5cf0b5ff-74a9-48b2-9506-d7ba006cd58f">

which imho are not very relevant for the Frontend, and required only when you need to do some template evaluation or what ever in the backend.

there are many examples for this enum attribute:

<img width="597" alt="Scherm­afbeelding 2023-12-04 om 10 09 55" src="https://github.com/NemesisRE/kiosk-mode/assets/33354141/73be5cba-7ccb-48c2-875e-c902a4a024f8">
<img width="608" alt="Scherm­afbeelding 2023-12-04 om 10 10 10" src="https://github.com/NemesisRE/kiosk-mode/assets/33354141/3bfde101-8a80-4b1f-bfa3-300a09fc4aba">
<img width="596" alt="Scherm­afbeelding 2023-12-04 om 10 10 32" src="https://github.com/NemesisRE/kiosk-mode/assets/33354141/d1703738-8bac-44c9-8b95-58070671029b">

being able to keep those out of the Frontend would clean up those cards a lot. 

I do realize it requires some magic to not only not show these attributes, but also Hide the attributes box completely, if it were the only attribute....

see our earlier discussion on that subject

**Reply by elchininet** [reply]
Hi @Mariusthvdb,

Thanks for the screenshots, everything is clear.

In my personal opinion, the content of the attributes is something that most users would want to hide altogether in Kiosk-devices (Mac addresses, hosts IP, serial numbers, events ids, etc), that is why there is an option that allows one to hide them. But hiding specific items of the attributes, not only is out of the scope of the plugin, but it will open a Pandora box because there are tons of attributes and each integration can add any number of them.

But that is something that could be added to [the Kiosk-mode complements](https://github.com/NemesisRE/kiosk-mode/blob/master/KIOSK-MODE-COMPLEMENTS.md) though, maybe addressing the issues that make your current implementation not up to a the modern dashboard implementation.

@NemesisRE has the last call on this one, but this is just my opinion on this matter.

Regards

**Reply by Mariusthvdb** [reply]
sure, thanks for your in-depth reply, as always

my thought would have been to set an option for hide_attribute, and let the user add a single or list of attributes. Not that Kiosk mode would set a predefined attribute.

In core HA, there is a 'filtered attributes' list, and my hope was to add the manual addition to that list of filtered attributes.

I believe this can not be done with Card-mod, since that only manipulates frontend

see: https://github.com/home-assistant/frontend/blob/dev/src/components/ha-attributes.ts

and  [extra_filters](https://github.com/home-assistant/frontend/blob/6d0eb059544608e3aee2e93315a9b4fb17dc4e74/src/components/ha-attributes.ts#L24)

whihc in the end all uses https://github.com/home-assistant/frontend/blob/aa4acc65724964339014b4a9724cc1e4182342ec/src/data/entity_attributes.ts#L1 as list of filtered attributes.

seems that instead of doing what my update_more_info currently does:

```
    if (t.length) {
      let e;
      for (const node of t) {
        const o = node.getElementsByClassName("key")[0];
        if (o.innerText.toLowerCase().trim() == "hide attributes") {
// make compatible for both 2023.8 and 2023.9
          const valueContainer = o.parentNode.getElementsByClassName("value")[0];
          const haAttributeValue = valueContainer.querySelector('ha-attribute-value');
          const text = haAttributeValue
            ? haAttributeValue.shadowRoot.textContent
            : valueContainer.innerText;
          e = text
            .split(",")
            .map((item) => item.replace("_", " ").trim());
          e.push("hide attributes");
        }
      }
      for (const node of t) {
        const o = node.getElementsByClassName("key")[0];
        if (
          e.includes(o.innerText.toLowerCase().trim()) ||
          e.includes("all")
        ) {
          o.parentNode.style.display = "none";
        }
```

is we could set that extra_filter like core does it (example here of the person entity)
```
<ha-attributes
--
.hass=${this.hass}
.stateObj=${this.stateObj}
extra-filters="id,user_id,editable,device_trackers"
></ha-attributes>
`;
```
and then even do that generically, and not only for a single domain.

but I am no dev so dont know how to reliably do so ...

**Reply by NemesisRE** [reply]
@elchininet is doing a damn good job at hiding almost every element of HA but at some point we have to think about our goal and the goal was "kiosk mode" which means "No touching, only look" originally that was only header and sidebar now we are more and more going towards RBAC and that could be fine if we would switch focus.

I for one think we should not switch focus but as long the code is easily manageable and new HA versions or different cards don't break the code every time we could add even RBAC stuff. We should also keep in mind that this just a lovelace mod and no real access control system if some one thinks this makes HA secure they would be misguided.

[the Kiosk-mode complements](https://github.com/NemesisRE/kiosk-mode/blob/master/KIOSK-MODE-COMPLEMENTS.md) on the other hand can be extended with everything, what could help.

**Reply by elchininet** [reply]
>I believe this can not be done with Card-mod, since that only manipulates frontend

Hi @Mariusthvdb,

What is the part that you believe cannot be done with `card-mod`?

Because that is precisely what `kiosk-mode` does, we have access only to the DOM and we only manipulate front-end stuff 🙂

**Comment by Mariusthvdb** [comment]
> @elchininet is doing a damn good job at hiding almost every element of HA

I couldnt agree more! amazing indeed.

> What is the part that you believe cannot be done with card-mod?

I don't think card_mod can add attributes to the extra-filters as linked in the files above? If so, it would be amazing, but I've never been able to manage that...

maybe a tiny custom integration/custom card could do that, it would probably only require a few lines of code.
But as said, I cant make that lacking the dev skill set for such an undertaking.

**Reply by elchininet** [reply]
I was not referring to add extra-filters (this is something that can not be done from a front-end code) but to be able to hide specific attributes from the select.

**Reply by Mariusthvdb** [reply]
hehe, we're going in circles here.... 

the reason this shouldn't be done from the frontend is that if it renders no more attributes to list in that dropdown, it will still show the dropdown, albeit empty. 

That is exactly what my own custom-ui/update more-info does.

My aim is to change that way of hiding those attributes, and add them in the backend to the extra-filters. As that will be the only truly functional way of hiding selective attributes, and not showing the box when empty

Bram once gave me this to do it:
```
const jouwFilters = "attribute1, attribute2, attribute3"
// find the right queryselector here
const haAttributes = querySelector("ha-attributes");
haAttributes.extraFilters = haAttributes.extraFilters + (haAttributes.extraFilters.length ? "," : "") + jouwFilters;
```

but as said, I couldnt make it happen

**Reply by elchininet** [reply]
>My aim is to change that way of hiding those attributes, and add them in the backend to the extra-filters. As that will be the only truly functional way of hiding selective attributes, and not showing the box when empty

First thing is to test if we can get the extra filters in that way.
I can test it to see if that is possible if not we are talking merely about "it would be good to have".

**Reply by elchininet** [reply]
>maybe a tiny custom integration/custom card could do that, it would probably only require a few lines of code.
>But as said, I cant make that lacking the dev skill set for such an undertaking.

I‘ll try to check if that is possible, if it is possible I see this more in the context of something specific to do that task (maybe together with the code that you already separated from custom-ui) and I can help you to make it works so it can be published in HACS.

**Reply by Mariusthvdb** [reply]
some googling and breaking down my own custom-ui (and copying the main structure) results in this code which still does not function as hoped for but throws n errors either.... 
```

const Name = "Custom-ui-filter-attributes";
const Version = "20231207";
const Description = "filter additional attributes from dropdown";
const Url = "https://github.com/Mariusthvdb/custom-ui-filter-attributes";
console.info(
  `%c  ${Name}  \n%c  Version ${Version} ${Description}`,
  "color: gold; font-weight: bold; background: black",
  "color: white; font-weight: bold; background: steelblue"
);
window.customUI = {
  installCustomFilters() {
    const CustomFilters = "options, icon_color";

    // Find the ha-attributes element
    const haAttributes = document.querySelector("ha-attributes");

    // Check if the element is found before accessing its properties
    if (haAttributes) {
      // Check if the extraFilters property is present before updating it
      if ('extraFilters' in haAttributes) {
        // Update the extraFilters property
        haAttributes.extraFilters += (haAttributes.extraFilters.length ? "," : "") + CustomFilters;
      } else {
        console.error("extraFilters property not found in ha-attributes");
      }
    } else {
      console.error("ha-attributes not found");
    }
  },
  installClassHooks() {
    window.customUI.installCustomFilters();
  },
  init() {
    window.customUI.initDone = true;
    window.customUI.installClassHooks();
    window.CUSTOM_UI_LIST = window.CUSTOM_UI_LIST || [];
    window.CUSTOM_UI_LIST.push({
      name: Name,
      version: `${Version} ${Description}`,
      url: Url
    });
  },
};

// Ensure the DOM is fully loaded before running the script
document.addEventListener("DOMContentLoaded", function() {
  window.customUI.init();
});
```

maybe it saves you the trouble of creating the structure to test with. I did add some tracing, but even that doesnt show, so a bit puzzled what to do next

**Reply by elchininet** [reply]
Hi @Mariusthvdb,

I was testing and yes, the `ha-attributes` lit element has an `extra-filters` property and it can be used to to send a list of filters to this element. I have tested with my HP printer that have several attributes:

<img width="1445" alt="image" src="https://github.com/NemesisRE/kiosk-mode/assets/3728220/2d77cfd7-8c81-4798-86f5-941a4bdd8fe8">

And this is what is rendered:

<img width="601" alt="image" src="https://github.com/NemesisRE/kiosk-mode/assets/3728220/fe5b2a3e-76c3-4b7c-aa07-fe24626fd66c">

If I send to the `ha-attributes`

```javascript
$0.extraFilters = 'marker_high_level,marker_low_level';
```

This is what is displayed:

<img width="596" alt="image" src="https://github.com/NemesisRE/kiosk-mode/assets/3728220/5fa71258-3dba-407a-b02e-79535050c81a">

Looking at your code it should do the work, what is failing is the selector:

```javascript
const haAttributes = document.querySelector("ha-attributes");
```

This will return `null` because the `ha-attributes` element is deep in the DOM tree and after several shadowDoms so it should log in the console:

>ha-attributes not found

This is easy to fix, and you can use the [home-assistant-query-selector](https://github.com/elchininet/home-assistant-query-selector) package that I prepared for this because it facilitates the work querying for Home Assistant DOM elements.

Let me know if you need some help with that and I can prepare an initial code for you so you can test it deeply in your environment.

Regards

**Reply by Mariusthvdb** [reply]
o this is getting to be wonderful ! amazing indeed.

could you test what happens if you filter all attributes showing on that printer? My hope is, with this technique, the length of that list is 0, so there will be no dropdown box for the attributes ...?

As for the Dom Path, I need to study that, but fear it will be simpler if you do that on your machine (the path should be identical for all of us? I wouldn't know where to start even installing that package.

**Reply by elchininet** [reply]
>could you test what happens if you filter all attributes showing on that printer? My hope is, with this technique, the length of that list is 0, so there will be no dropdown box for the attributes ...?

That‘s right. If all the attributes are added to `extraFilters`, the `ha-attributes` element is not shown.

>As for the Dom Path, I need to study that, but fear it will be simpler if you do that on your machine (the path should be identical for all of us? I wouldn't know where to start even installing that package.

Do not worry then, I‘ll try to create a `JavaScript` file for you, so you can test it in your installation.

**Reply by elchininet** [reply]
Hi @Mariusthvdb,
I pushed the code [here](https://github.com/elchininet/custom-ui-filter-attributes). Test the `custom-ui-filter-attributes.js` file (the one in the root, not the one inside `src`). If it works well, you can move all the code to your repo (it is basically the source file and the config to compile the final one).
Regards

**Reply by Mariusthvdb** [reply]
wow, I do believe it works as we had hoped for! 
on my season entity, no more options, and indeed, no drop-down attributes box

no more icon_color anywhere.!

this is truly great, I now have to figure out what you did ;-)

btw I had already opened a repo at https://github.com/Mariusthvdb/custom-attributes but if you like, please keep it at yours ofc course.

I have 2 questions:
1) since we hardcode filter 'options' now, this seems to also filter out the options on an input select. That is undesirable  as far as I am concerned, though nit entirely unexpected.... I only wished to filter out the options of enum device_class.

2) directly related to hard coding the filters, Id love to be able to set them via customize in yaml. that way we can do things like:

```
homeassistant:
  customize
    '*.*':
      filter_attributes:
        - icon_color

  customize_domain:
     person:
       filter_attributes:
         - source
```

not sure how to get to the enum values, without touching the (input_)selects though. Maybe you would know of a way?

Ill play with the for now, and try to find the correct spot where you set the current filters so we have a thoutough idea of the functionality. (I'll even re-install the former custom-ui to see if it can filter the templates too! that would be awesome

thanks for now, hope you had a bit of fun too

___update___

Yes, it also filter the `templates` attribute we set in custom-ui. so that makes it even better.

**Reply by elchininet** [reply]
> I now have to figure out what you did ;-)

Just go to `src/custom-ui-filter-attributes.js` and you can check the code.

>btw I had already opened a repo at https://github.com/Mariusthvdb/custom-attributes but if you like, please keep it at yours ofc course.

Move the files to your repo, I created that repo just for make a quick set up so you could test the code.

>since we hardcode filter 'options' now, this seems to also filter out the options on an input select. That is undesirable as far as I am concerned, though nit entirely unexpected.... I only wished to filter out the options of enum device_class.

You need to explain this a bit (we can do it later in a discussion in your repo), and I can recommend you a solution, but I need to understand what you are searching for, I addded `options` just because it was in your original code.

>directly related to hard coding the filters, Id love to be able to set them via customize in yaml. that way we can do things like.

Configuration is possible, but a front-end configuration in a lovelace dashboard (like the one in `kiosk-mode` or `keep-texts-in-tabs`), but if you are referring to any back-end yaml config, we cannot read those.

>not sure how to get to the enum values, without touching the (input_)selects though. Maybe you would know of a way?

I don‘t get the question, let‘s discuss it in your repo, just move the code there and we can discuss in the correct scope.

>thanks for now, hope you had a bit of fun too

Welcome, glad to help 👍

**Reply by Mariusthvdb** [reply]
done, I copied the files an all is set at https://github.com/Mariusthvdb/custom-attributes
if you wouldn't mind I could share the site with you, make you a contributor?

you could also move this discussion maybe to he other repo, so your repo here can be cleansed a bit ;-)

**Reply by elchininet** [reply]
You can add me as a contributor, I wouldn't mind.
I‘ll move the discussion to yours.

>Edit: I cannot do it, it says "Move this discussion to another repository owned by Steven "NemesisRE" Kurz.".

**Reply by Mariusthvdb** [reply]
that is a pity. I'll see if I can copy some of the more important aspects into some background document then, or discussion.
meantime you've been made a contributor ;-)

**Comment by Mariusthvdb** [comment]
We should probably close this discussion and refer to https://github.com/Mariusthvdb/custom-attributes

**Reply by elchininet** [reply]
Makes sense 👍🏼
